internal/charmstore/store.go

```diff
@@ -493,19 +525,109 @@ func (s *Store) FindBestEntity(url *charm.URL, fields map[string]int) (*mongodoc
+		}
+		if channel == DevelopmentChannel && !entity.Development {
+			return nil, errgo.WithCausef(nil, params.ErrNotFound, "no matching charm or bundle for %s", url)
+		}
+		return entity, nil
+	}
+	switch channel {
+	case StableChannel, UnspecifiedChannel:
+		return s.findStableEntity(url, fields)
+	case DevelopmentChannel:
+		return s.findDevelopmentEntity(url, fields)
+	case UnpublishedChannel:
+		return s.findUnpublishedEntity(url, fields)
+	default:
+		return nil, errgo.WithCausef(nil, params.ErrNotFound, "no matching charm or bundle for %s", url)
+	}
+}
+
+// findStableEntity attempts to find an entity on the stable channel. The
+// base entity for URL is retrieved and the series with the best match to
+// URL.Series is used as the resolved entity.
+func (s *Store) findStableEntity(url *charm.URL, fields map[string]int) (*mongodoc.Entity, error) {
+	baseEntity, err := s.FindBaseEntity(url, map[string]int{
+		"_id":          1,
+		"stableseries": 1,
+	})
+	if errgo.Cause(err) == params.ErrNotFound {
+		return nil, errgo.WithCausef(nil, params.ErrNotFound, "no matching charm or bundle for %s", url)
+	} else if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	entityURL := baseEntity.StableSeries[url.Series]
+	if entityURL == nil && url.Series == "" {
+		for _, u := range baseEntity.StableSeries {
+			if entityURL == nil || seriesScore[u.Series] > seriesScore[entityURL.Series] {
+				u := u
+				entityURL = u
+			}
+		}
+	}
+	if entityURL == nil {
+		return nil, errgo.WithCausef(nil, params.ErrNotFound, "no matching charm or bundle for %s", url)
+	}
+	return s.findSingleEntity(entityURL, fields)
+}
+
+// findDevelopmentEntity attempts to find an entity on the development
+// channel. The base entity for URL is retrieved and the series with the
+// best match to URL.Series is used as the resolved entity.
+func (s *Store) findDevelopmentEntity(url *charm.URL, fields map[string]int) (*mongodoc.Entity, error) {
+	baseEntity, err := s.FindBaseEntity(url, map[string]int{
+		"_id":               1,
+		"developmentseries": 1,
+	})
+	if errgo.Cause(err) == params.ErrNotFound {
+		return nil, errgo.WithCausef(nil, params.ErrNotFound, "no matching charm or bundle for %s", url)
+	} else if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	entityURL := baseEntity.DevelopmentSeries[url.Series]
+	if entityURL == nil && url.Series == "" {
+		for _, u := range baseEntity.DevelopmentSeries {
+			if entityURL == nil || seriesScore[u.Series] > seriesScore[entityURL.Series] {
+				u := u
+				entityURL = u
+			}
+		}
+	}
+	if entityURL == nil {
+		return nil, errgo.WithCausef(nil, params.ErrNotFound, "no matching charm or bundle for %s", url)
+	}
+	return s.findSingleEntity(entityURL, fields)
+}
 
+// findUnpublishedEntity attempts to find an entity on the unpublished
+// channel. This searches all entities in the store for the best match to
+// the URL.
+func (s *Store) findUnpublishedEntity(url *charm.URL, fields map[string]int) (*mongodoc.Entity, error) {
```

Comment:
I wonder if we should just delete this logic. It's really just here for the tests AFAICS. Still, I guess it's probably still useful.